feat: Change Invocation.format to Invoke

gauron99 opened the pull request:
<!-- Thanks for sending a pull request! -->

# Changes

- Change Invocation structure from f.Invocation.format -> f.Invoke
- update template files & tests
- create migration 0.35.0 from previous versions & add tests

/kind enhancement
/fixes #1280

lkingland commented on client.go:
```diff
@@ -879,7 +879,10 @@ func (c *Client) Invoke(ctx context.Context, root string, target string, m Invok
 		return
 	}
-
+	// if invoke is empty, use default 'http'
+	if f.Invoke == "" {
+		f.Invoke = DefaultInvocationFormat
+	}
```
I would go a step further and remove this condition from here because the `invoke` method already takes care of this logic.  It might just need to have the defaulting moved up to the very top of the function (before the verbose print statement).  The `invoke` method can expect to default to invoking via HTTP if not given an invocation method explicitly on the function or the message.

lkingland commented on function_migrations.go:
```diff
@@ -81,6 +81,7 @@ var migrations = []migration{
 	{"0.23.0", migrateToBuilderImages},
 	{"0.25.0", migrateToSpecVersion},
 	{"0.34.0", migrateToSpecsStructure},
+	{"0.35.0", migrateToInvokeStructure},
```
```suggestion
	{"0.35.0", migrateFromInvokeStructure},
```
Maybe?

lkingland commented on function_migrations.go:
```diff
@@ -340,3 +362,14 @@ type migrateToSpecVersion_previousFunction struct {
+
+// Functions prior to 0.35.0 will have Invocation.Format instead of Invoke
+type migrateToInvokeStructure_previousFunction struct {
+	Invocation Invocation `yaml:"invocation,omitempty"`
```
We could rename this struct something like migrateFromInvokeStructure_invocation. If the `yaml:"invocation"` annotation on the struct field stays the same, it should have no effect on the migration.
That way, this struct is no longer part of the package's exported Godocs, and will not get in the way if we ever end up adding the structure back.

lkingland commented on template.go:
```diff
@@ -52,8 +52,8 @@ func (t template) Write(ctx context.Context, f *Function) error {
 	}
-	if f.Invocation.Format == "" {
-		f.Invocation.Format = t.config.Invocation.Format
+	if f.Invoke == "" && t.config.Invoke != DefaultInvokeFormat {
```
```suggestion
	if f.Invoke == "" && t.config.Invoke != "http" {
```
I would change this to the actual value, since that is "locked in time" as being the value pre-migration-35, whereas it is possible (though very unlikely) that we could change DefaultInvokeFormat in the future

lkingland reviewed (commented):
Looks good!  I just made a few naming suggestions etc.

gauron99 commented on function_migrations.go:
```diff
@@ -282,6 +283,32 @@ func migrateToSpecsStructure(f1 Function, m migration) (Function, error) {
+		return f1, errors.New("migration 'migrateFromInvokeStructure' error: " + err.Error())
+	}
+
+	if f0.Invocation.Format != "" && f0.Invocation.Format != DefaultInvocationFormat {
```
I was wondering - Should i change this Default format to "locked in time" "http" string as well?

matejvasek commented:
@gauron99 you will need to regenerate `zz_filesystem_generated.go` because I modified it in another PR.

lkingland commented on function_migrations.go:
```diff
@@ -340,3 +367,14 @@ type migrateToSpecVersion_previousFunction struct {
+
+// (Defined only for previous versions migration)
+// Invocation defines hints on how to accomplish a function invocation.
+type Invocation struct {
```
```suggestion
type migrateFromInvokeStructure_invocation struct {
```

lkingland commented on function_migrations.go:
```diff
@@ -340,3 +367,14 @@ type migrateToSpecVersion_previousFunction struct {
+}
+
+// Functions prior to 0.35.0 will have Invocation.Format instead of Invoke
+type migrateFromInvokeStructure_invocation struct {
```
```suggestion
type migrateFromInvokeStructure struct {
```

lkingland commented on function_migrations.go:
```diff
@@ -282,6 +283,32 @@ func migrateToSpecsStructure(f1 Function, m migration) (Function, error) {
+	if err != nil {
+		return f1, errors.New("migration 'migrateFromInvokeStructure' error: " + err.Error())
+	}
+	f0 := migrateFromInvokeStructure_invocation{}
```
```suggestion
	f0 := migrateFromInvokeStructure{}
```

lkingland commented on function_migrations.go:
```diff
@@ -340,3 +367,14 @@ type migrateToSpecVersion_previousFunction struct {
+
+// Functions prior to 0.35.0 will have Invocation.Format instead of Invoke
+type migrateFromInvokeStructure_invocation struct {
+	Invocation Invocation `yaml:"invocation,omitempty"`
```
```suggestion
	Invocation migrateFromInvokeStructure_invocation `yaml:"invocation,omitempty"`
```

lkingland commented on function_migrations_unit_test.go:
```diff
@@ -170,3 +170,30 @@ func TestMigrateToSpecs(t *testing.T) {
 
 }
+
+// TestMigrateToInvokeStructure tests that migration from f.Invocation.Format to
+// f.Invoke works
+func TestMigrateToInvokeStructure(t *testing.T) {
```
```suggestion
// TestMigrateFromInvokeStructure tests that migration from f.Invocation.Format to
// f.Invoke works
func TestMigrateFromInvokeStructure(t *testing.T) {
```

lkingland reviewed (commented):
Looks good.  I clarified the suggestion to rename the `Invoke` struct via suggestions herein:

gauron99 commented:
I see now what you meant. I changed the names back and added `_previousFunction` for the `struct` because otherwise the name was with conflict with the`func` itself

lance commented:
@gauron99 looks like you need to regenerate zz_filesystem_generated.go - sorry about that!

zroubalik reviewed (approved):
/lgtm

zroubalik commented:
/override "Func E2E OnCluster RT Test"

zroubalik commented:
/override "On Cluster RT Test (1.18.x, ubuntu-latest)"
